v2/internal/identity/credential_provider.go

```diff
@@ -0,0 +1,260 @@
+	}
+
+	if c.globalCredential == nil {
+		return nil, errors.New("global credential not configured, you must use either namespaced or per-resource credentials")
+	}
+	// If not found, default is the global credential
+	return c.globalCredential, nil
+}
+
+func (c *credentialProvider) getCredentialFromNamespaceSecret(ctx context.Context, namespace string) (*Credential, error) {
+	secretNamespacedName := types.NamespacedName{Namespace: namespace, Name: NamespacedSecretName}
+	cred, err := c.getCredentialFromSecret(ctx, secretNamespacedName)
+	if err != nil {
+		var secretNotFound *core.SecretNotFound
+		if errors.As(err, &secretNotFound) {
+			return nil, nil // Not finding this secret is allowed, allow caller to proceed to higher scope secret
+		}
+		if apierrors.IsNotFound(err) {
+
+		}
+		return nil, err
+	}
+
+	return cred, nil
+}
+
+func (c *credentialProvider) getCredentialFromAnnotation(ctx context.Context, obj genruntime.MetaObject, annotation string) (*Credential, error) {
+	credentialFrom, ok := obj.GetAnnotations()[annotation]
+	if !ok {
+		return nil, nil
+	}
+
+	// Disallow credentials with `/` in their credentialFrom
+	if strings.Contains(credentialFrom, "/") {
+		err := errors.Errorf("%s cannot contain '/'. Secret must be in same namespace as resource.", annotation)
+		namespacedName := types.NamespacedName{
+			Namespace: obj.GetNamespace(),
+			Name:      credentialFrom,
+		}
+		return nil, core.NewSecretNotFoundError(namespacedName, err)
+	}
+
+	// annotation exists, use specified secret
+	secretNamespacedName := getSecretNameFromAnnotation(credentialFrom, obj.GetNamespace())
+	return c.getCredentialFromSecret(ctx, secretNamespacedName)
+}
+
+func (c *credentialProvider) getCredentialFromSecret(ctx context.Context, secretNamespacedName types.NamespacedName) (*Credential, error) {
+	secret, err := c.getSecret(ctx, secretNamespacedName.Namespace, secretNamespacedName.Name)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil, core.NewSecretNotFoundError(secretNamespacedName, errors.Wrapf(err, "credential secret not found"))
+		}
+		return nil, err
+	}
+
+	cred, err := c.newCredentialFromSecret(secret)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to get credential %q", secretNamespacedName.String())
+	}
+	return cred, nil
+}
+
+func (c *credentialProvider) newCredentialFromSecret(secret *v1.Secret) (*Credential, error) {
+	var errs []error
+
+	nsName := types.NamespacedName{Namespace: secret.GetNamespace(), Name: secret.GetName()}
+
+	subscriptionID, ok := secret.Data[config.SubscriptionIDVar]
+	if !ok {
+		err := core.NewSecretNotFoundError(nsName, errors.Errorf("credential Secret %q does not contain key %q", nsName.String(), config.SubscriptionIDVar))
+		errs = append(errs, err)
+	}
+
+	tenantID, ok := secret.Data[config.TenantIDVar]
+	if !ok {
+		err := core.NewSecretNotFoundError(nsName, errors.Errorf("credential Secret %q does not contain key %q", nsName.String(), config.TenantIDVar))
```

Comment:
Suggest formatting for readability

```suggestion
		err := core.NewSecretNotFoundError(
		    nsName, 
		    errors.Errorf(
		        "credential Secret %q does not contain key %q", 
    		        nsName.String(), 
    		        config.TenantIDVar))
```

Reply:
Done